### ISSUE TITLE
[LSP4MP] Diagnostics regarding properties file do not update instantly

### ISSUE THREAD

**Issue by evie-lau** [OpenLiberty/liberty-tools-intellij#328]
As we share a codebase with [intellij-quarkus](https://github.com/redhat-developer/intellij-quarkus), this issue is present there too. The VSCode extension `Tools for MicroProfile` does not have this same behavior.

When updating a MP config properties file, it can take up to a minute for the changes to be reflected in memory. During this time, any diagnostics refresh on the java file referencing the properties file will be using an outdated version of the properties file.

Even when the memory copy of the properties file gets updated, `publishDiagnostics` isn't triggered automatically, and something else has to trigger it to update the diagnostics and refresh the problems.

This is especially noticeable when using a quickfix to add a property into the properties file, because the diagnostic will stick around until after the system file gets re-read, and the diagnostics are triggered again.

In summary: to have the MP diagnostics accurately reflect a changed properties file, you must...

1. wait up to 60 seconds for the in-memory copy to read from the file in system
2. manually trigger a `publishDiagnostic` on the java file you are viewing (changing or closing/reopening the file)

**Comment by mrglavas** [comment]
@evie-lau Do you happen to have a test case I could use to reproduce this?

**Comment by evie-lau** [comment]
After some more debugging with @mrglavas, here are some things we observed.
Update: it is not a timing issue... it requires switching windows on the OS.

When applying a quickfix or manual change to modify microprofile-config.properties:
- nothing will ever evict MP config cache until switching windows
- nothing will ever loadConfigSource until making changes to the java file, AFTER the config cache has been evicted

loadConfigSource is how we pull in the contents of the config file, but it only runs when the cache has been evicted (is null). The cache does not get nulled until switching out of the IDE window, so diagnostics will never have an updated copy of the config file to work with until this happens.

Evict call comes from MicroProfileProjectListener (implements ModuleListener and BulkFileListener)
loadConfigSources comes from PsiMicroProfileProject

**Comment by mrglavas** [comment]
@evie-lau When we were debugging yesterday were you saving the files? The response to this post on JetBrains' forums: https://intellij-support.jetbrains.com/hc/en-us/community/posts/360006715420-Virtual-File-changes-only-occur-when-clicking-off-Intellij suggests that events are only fired when changes are saved on disk. Merely typing in the editor isn't enough. Perhaps the quickfix for `microprofile-config.properties` is only queuing up a file write and switching windows forces the in memory document to be written to the disk.

**Comment by evie-lau** [comment]
Ah, I had thought that IntelliJ just automatically saved and didn't allow the user to manually save. I didn't know the mechanism behind how/when it did so. That does seem to match up - wondering if we can force a disk write for the mp config files in particular (whether edited via quickfix or manual change).

**Comment by angelozerr** [comment]
@evie-lau I discovered your issue and IMHO, we need to have the same code base for MicroProfile if we want to provide a MicroProfile IJ plugin that we could extend for instance with Quarkus, like we have with vscode-microprofile / vscode-quarkus.
